### DISCUSSION TITLE
Using "using" to map auto-implemented properties to a value source

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#2291] [Language Ideas]
I suggest to use `using` to map auto-implemented properties to a variable, a filed or an indexer... etc.
Syntax:
```C#
public int NewYork {get; set; } using this["NewYork"];
```

Which is a shortcut for:
```C#
public int NewYork
{
   get => this["NewYork"];
   set => this["NewYork"] = value;
}
```
This also allows:
```C#
   public int Age {get; set; } using _age = 10;
```

Which is a shortcut for:
```C#
int _age = 10;
public int Age
{
   get => _age;
   set => _age = value;
}
```

which gives you access to the backing field _age.

Most of the time, we write a full property, just to link it directly to an item of a collection or a property of a control. Example:
```C#
public int Foo
{
   get => x / 10;
   set => x = value * 10;
}
```

Can be reduced to:
```C#
public int Foo {get; set;} using x/10;
```

It will need a little work to calculate the correct formula of the setter, as 
`x/10 = value` => `x = 'value * 10`
the same for using x - 2:
`x-2 = value` => `x = 'value + 2`
and Math.Pow(x, 2)
`Math.Pow(x, 2) = value` => `x = 'Math.Pow(value , 1/2)`
We can set some simple roles like this, and any more complex relations should be written in a full property syntax.

**Comment by john-h-k** [comment]
So you just want 2 way property `=>` syntax without explicitly having a getter and setter?

**Comment by ghost** [comment]
@johnkellyoxford 
Yes. Most of the time, we write a full property, just to link it directly to an item of a collection or a property of a control. Example:
```C#
public int Foo
{
   get => x / 10;
   set => x = value * 10;
}
```

Can be reduced to:
```C#
public int Foo {get; set;} using x/10;
```

It will need a little work to calculate the correct formula of the setter, as 
`x/10 = value` => `x = 'value * 10`
the same for using x - 2:
`x-2 = value` => `x = 'value + 2`
and Math.Pow(x, 2)
`Math.Pow(x, 2) = value` => `x = 'Math.Pow(value , 1/2)`
We can set some simple roles like this, and any more complex relations should be written in a full property syntax.

**Comment by TPFInferno** [comment]
I don't like applying a new meaning to the using keyword.  From the syntax above it just confuses me and makes me think that your source implements IDisposable, which would make it impossible to infer when the Dispose method would even be called.

Here is a syntax I think would look a lot better:

```cs
public int NewYork { get; set; } => this["NewYork"];
```

The get and set would need to be in the declaration in order for the compiler to infer that the property also sets.

But, honestly I don't think there is much value in this. Nothing substantial is solved or alleviated from developers with this kind of change.

**Comment by john-h-k** [comment]
I really don't think this is very useful

But if it was, I'd go with a syntax something like:

* `<=>` (as opposed to `=>`)
* Auto property + =>, e.g `{ get; set; } => otherInt`

 Adding a new meaning to using is just conf **using**
